### PR TITLE
RAINCATCH-1071 - Image improvements

### PR DESCRIPTION
**Pull request by wtrocki**
## What was done

- [x] Minor image improvements
- [x] Remove custom run step
- [x] Image no longer needs to be run as root 
- [x] enable and test openshift integration
- [x] publish image

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -14,35 +14,18 @@ ENV KEYCLOAK_BIN_FOLDER $KEYCLOAK_USER_HOME_FOLDER/keycloak-$KEYCLOAK_VERSION/bi
...
-RUN sh $KEYCLOAK_BIN_FOLDER/add-user-keycloak.sh -u $KEYCLOAK_ADMIN_USERNAME -p $KEYCLOAK_ADMIN_PASSWORD
-
-# Add the realm file to create the raincatcher realm
-USER root
-COPY ./data_files/raincatcher-realm.json $DATA_FILE_FOLDER
-USER $KEYCLOAK_SYSTEM_USER
+COPY ./scripts $SCRIPTS_FOLDER
+COPY ./data_files $DATA_FILE_FOLDER
+RUN ls $DATA_FILE_FOLDER && ls $SCRIPTS_FOLDER
+RUN chmod +x $SCRIPTS_FOLDER/*
```
Comment (twnolan) [inline Dockerfile]: I had a couple of issues when trying to run this images

**Issue 1** Building failed at the following step. 

```
Step 21 : RUN chmod +x $SCRIPTS_FOLDER/*
 ---> Running in 6dfe62cdb371
chmod: /home/keycloak/scripts/docker-entrypoint.sh: Operation not permitted
The command '/bin/sh -c chmod +x $SCRIPTS_FOLDER/*' returned a non-zero code: 1Step 21 : RUN chmod +x $SCRIPTS_FOLDER/*
 ---> Running in 6dfe62cdb371
chmod: /home/keycloak/scripts/docker-entrypoint.sh: Operation not permitted
The command '/bin/sh -c chmod +x $SCRIPTS_FOLDER/*' returned a non-zero code: 1
```
This is due to COPY and ADD both creating new files with UID and GID of 0 - https://docs.docker.com/engine/reference/builder/#add https://docs.docker.com/engine/reference/builder/#copy

If you change the previous line to:

`RUN ls -la $DATA_FILE_FOLDER && ls -la $SCRIPTS_FOLDER`

You will get the following output of directories owners and permissions:

```
Step 20 : RUN ls -la $DATA_FILE_FOLDER && ls -la $SCRIPTS_FOLDER
 ---> Running in 736f812acc09
total 12
drwxr-sr-x    3 root     root            50 Jul 31 11:54 .
drwxr-sr-x    5 keycloak nogroup         67 Jul 31 11:54 ..
-rw-r--r--    1 root     root          9878 Jul 31 11:15 raincatcher-realm.json
drwxr-xr-x    3 root     root            40 Jul 31 11:15 themes
total 4
drwxr-sr-x    2 root     root            34 Jul 31 11:54 .
drwxr-sr-x    5 keycloak nogroup         67 Jul 31 11:54 ..
-rw-r--r--    1 root     root           828 Jul 31 11:41 docker-entrypoint.sh
```
The chmod command will not run in this case unless its run by the root user

**Issue 2** No admin user created

```
-# Create admin user for keycloak
 -RUN sh $KEYCLOAK_BIN_FOLDER/add-user-keycloak.sh -u $KEYCLOAK_ADMIN_USERNAME -p $KEYCLOAK_ADMIN_PASSWORD
```

This was removed but is required as it generates our admin user so we can login.

File: README.md [hunk 2]
```diff
@@ -1,17 +1,24 @@
...
-## Running
+
+> Note: We do not use official keycloak image as base. Image is built on the alpine base to provide a small image to run tests
+against. The image size is ~309 MB as opposed to the JBoss/Keycloak image which is ~640 MB.
+
+## Running image 
+
+Image is published to docker hub 
+
+    docker run feedhenry/raincatcher-keycloak
```
Comment (twnolan) [inline README.md]: `docker run -p 8080:8080 feedhenry-keycloak`

We will need to use port binding to access as localhost for demos and testing as well

File: README.md [hunk 3]
```diff
@@ -1,17 +1,24 @@
...
-docker image built on the alpine base to provide a small image to run tests
-against. The image size is ~309 MB as opposed to the JBoss/Keycloak image which
-is ~640 MB.
+The contents of this repository are used to create a Keycloak standalone server
+docker image. This image is designed and styled to work with RainCatcher demo solution.
 
-## Running
+
+> Note: We do not use official keycloak image as base. Image is built on the alpine base to provide a small image to run tests
+against. The image size is ~309 MB as opposed to the JBoss/Keycloak image which is ~640 MB.
```
Comment (twnolan) [inline README.md]: Image size is now 330.1 MB due to the themes being added

**Review by twnolan — DISMISSED** [review]
Some issues currently that I have addressed In comments below, but looks good. Note I couldn't login to check the data due to the second issue in the Dockerfile

**Comment by wtrocki** [comment]
@TommyJ1994 @twnolan - Image is working now.
Pushed version to dockerhub.
After logging I do not see Raincatcher them being created.

File: scripts/docker-entrypoint.sh [hunk 4]
```diff
@@ -1,4 +1,8 @@
 #!/bin/sh
+set -o errexit
 
-exec $KEYCLOAK_BIN_FOLDER/standalone.sh $@
+ls -al
```
Comment (tomjackman) [inline scripts/docker-entrypoint.sh]: can be removed?

**Review by tomjackman — COMMENTED** [review]
The theme folder/files are only prepared but they are not currently being used. This can be added later in a separate ticket.

File: scripts/realm-init.sh [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        installRealm
+    elif [ "$STAT" = "" ]; then 
+        sleep 5
+        echo "Keycloak is not running:`$STAT`. Trying again."
+        checkIfKeycloakRunning
+    fi
+}
+ 
+function installRealm {
+    echo "Installing RainCatcher realm into Keycloak Server"
```
Comment (tomjackman) [inline scripts/realm-init.sh]: "Importing the RainCatcher realm into the Keycloak Server"

File: scripts/realm-init.sh [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        sleep 10
+        installRealm
+    elif [ "$STAT" = "" ]; then 
+        sleep 5
+        echo "Keycloak is not running:`$STAT`. Trying again."
+        checkIfKeycloakRunning
+    fi
+}
+ 
+function installRealm {
```
Comment (tomjackman) [inline scripts/realm-init.sh]: `importRealm` might be better naming to keep it consistent with keycloak docs.

File: README.md [hunk 7]
```diff
@@ -1,42 +1,39 @@
...
-
-The name of the running container is "keycloak". Execute the following command
-to stop the running container:
-
-    docker stop keycloak
-
-## Restarting the stopped
-
-    docker start keycloak
+"Raincatcher" realm is visible along with the master realm.
```
Comment (tomjackman) [inline README.md]: Can we add back in the note from before too like so. ("The Raincatcher" realm is visible along with the master realm in the dropdown menu on the upper left corner of the page.) Someone new to Keycloak will only be provided with the misleading master client listing on server start, the quick sentence on how to switch realms would be helpful in my opinion as the realm switcher isn't obvious in the UI.

File: README.md [hunk 8]
```diff
@@ -1,42 +1,39 @@
...
+## Building image
 
 Execute the following commands to build and run the server:
 
-    chmod +x build_run_populate.sh
-    ./build_run_populate.sh
+    docker build -t feedhenry/raincatcher-keycloak .
 
 This will build the docker image, start the server with an admin user generated, and then populate the server
 with some test data specified from the data_files/raincatcher-realm.json file.
```
Comment (tomjackman) [inline README.md]: `data_files/raincatcher-realm.json` or **data_files/raincatcher-realm.json** perhaps.

**Comment by TommyJ1994** [comment]
In the last section of the README, the docs are referencing old files - "...... any running keycloak container, and run the build_run_populate.sh script again"

**Comment by tomjackman** [comment]
LGTM. Best for @twnolan to have final review.

File: README.md [hunk 9]
```diff
@@ -1,20 +1,30 @@
-Custom Keycloak Docker
+Raincatcher Demo Keycloak example. 
 ----------------------
 
-the contents of this repository are used to create a Keycloak standalone server
-docker image built on the alpine base to provide a small image to run tests
-against. The image size is ~309 MB as opposed to the JBoss/Keycloak image which
```
Comment (twnolan) [inline README.md]: Image size is now 344 MB, but saying its "approximately half the size" might be better than having to change a value after every change.

**Comment by wtrocki** [comment]
ping  @austincunningham for documentation changes.

**Review by twnolan — APPROVED** [review]
LGTM

**Pull request merged**